server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -57,39 +62,105 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
         let instant = Instant::now();
         let ctx = self.handler_ctx();
         let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
-        let handle = self.runtimes.read_runtime.spawn(async move {
-            let read_request = request.into_inner();
-            handle_stream_read(ctx, read_request).await
-        });
+
+        let request = request.into_inner();
+        let table_engine::remote::model::ReadRequest {
+            table,
+            read_request,
+        } = request.clone().try_into().box_err().context(ErrWithCause {
+            code: StatusCode::BadRequest,
+            msg: "fail to convert read request",
+        })?;
+
+        let request_key = RequestKey::new(
+            table.table,
+            read_request.predicate.clone(),
+            read_request.projected_schema.projection(),
+            read_request.order,
+        );
+
+        {
+            let mut dedup_map = self.dedup_map.write().await;
+            match dedup_map.get_notifiers(request_key.clone()) {
+                Some(notifiers) => {
+                    let txs = notifiers.get_txs();
+                    let mut notifiers = Notifiers::new(txs.clone());
+                    notifiers.add_tx(tx);
+                    dedup_map.add_notifiers(request_key.clone(), notifiers);
```

Comment:
```suggestion
                    notifiers.add_tx(tx);
```
It looks like the code is duplicated here.